ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -474,36 +484,90 @@ public Void visitFunctionCall(
         }
       }
 
-      final SqlType returnSchema = udfFactory.getFunction(argTypes).getReturnType(argTypes);
-      expressionTypeContext.setSqlType(returnSchema);
+      final KsqlScalarFunction function = udfFactory.getFunction(argTypesForFirstPass);
+
+      if (node.hasLambdaFunctionCallArguments()) {
+        final List<ParamType> paramTypes = function.parameters();
+        final Map<GenericType, SqlType> reservedGenerics = new HashMap<>();
+        final List<SqlArgument> argForReturnType = new ArrayList<>();
+        for (int i = 0; i < arguments.size(); i++) {
+          final Expression expression = arguments.get(i);
+          final ParamType parameter = paramTypes.get(i);
+          if (expression instanceof LambdaFunctionCall) {
+            final TypeContext childContext = expressionTypeContext.getCopy();
+
+            // the function returned from the UDF factory should have lambda
+            // at this index in the function arguments if there's a
+            // lambda node at this index in the function node argument list
+            if (!(parameter instanceof LambdaType)) {
+              throw new KsqlException("Error while processing lambda function. "
+                  + "This is most likely an internal error and a "
+                  + "Github issue should be filed for debugging.");
+            }
+
+            final LambdaType lambdaParameter = (LambdaType) parameter;
+            for (ParamType inputParam : lambdaParameter.inputTypes()) {
+              if (inputParam instanceof GenericType) {
+                final GenericType genericParam = (GenericType) inputParam;
+                if (!reservedGenerics.containsKey(genericParam)) {
+                  throw new KsqlException(
+                      String.format(
+                          "Could not verify type for generic %s.",
+                          genericParam.toString()));
+                }
+                childContext.addLambdaInputType(reservedGenerics.get(genericParam));
+              } else {
+                childContext.addLambdaInputType(
+                    SchemaConverters.functionToSqlConverter().toSqlType(inputParam));
+              }
+            }
+            final List<SqlType> sqlTypesForLambdaArgMapping =
+                new ArrayList<>(childContext.getLambdaInputTypes());
+            process(expression, childContext);
+
+            final SqlType resolvedArgType = childContext.getSqlType();
+            final SqlArgument lambdaArgument = SqlArgument.of(
+                SqlLambda.of(sqlTypesForLambdaArgMapping, resolvedArgType));
+
+            argForReturnType.add(lambdaArgument);
+            GenericsUtil.reserveGenerics(
+                parameter,
+                lambdaArgument,
+                reservedGenerics
+            );
+          } else {
+            argForReturnType.add(argTypesForFirstPass.get(i));
+            GenericsUtil.reserveGenerics(
```

Comment:
I don't know if we support lambdas in the middle of function declarations, but I think we'll want to process all non-lambda arguments before we process lambda arguments instead of doing them in the order they show up - otherwise we may not have the info we need to resolve the lambdas:
```
my_filter(x -> true, list)
```

if we don't support this, then we should fail hard if we see any non-lambda parameter passed in after a lambda paramter (which we might already be doing, I haven't looked at tests here)

Reply:
This logic should be fine. The parser only recognizes lambda functions at the end of a function https://github.com/confluentinc/ksql/blob/master/ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4#L300

We may want to change this in the future though if we want to support UDF with functions anywhere in the signature but I think this is low priority and should be a lower priority. 

I filed an issue to keep track of this https://github.com/confluentinc/ksql/issues/7107

Reply:
+1 from me - thanks for filing the ticket